File: app/src/main/java/com/github/mbarrben/moviedb/movies/MoviesRecyclerView.java
package com.github.mbarrben.moviedb.movies;

import android.content.Context;
import android.content.res.TypedArray;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.util.AttributeSet;

public class MoviesRecyclerView extends RecyclerView {

  private GridLayoutManager manager;
  private int columnWidth = -1;

  public MoviesRecyclerView(Context context, AttributeSet attrs) {
    super(context, attrs);
    init(context, attrs);
  }

  @Override
  protected void onMeasure(int widthSpec, int heightSpec) {
    super.onMeasure(widthSpec, heightSpec);

    if (columnWidth <= 0) {
      return;
    }

    setSpanCount();
  }

  private void init(Context context, AttributeSet attrs) {
    setHasFixedSize(true);
    addItemDecoration(new MovieItemDecoration(getContext()));
    initColumnsWidth(context, attrs);
    manager = new GridLayoutManager(getContext(), 1);
    setLayoutManager(manager);
  }

  private void initColumnsWidth(Context context, AttributeSet attrs) {
    if (attrs == null) {
      return;
    }

    int[] attrsArray = {
        android.R.attr.columnWidth
    };

    TypedArray array = context.obtainStyledAttributes(attrs, attrsArray);
    columnWidth = array.getDimensionPixelSize(0, -1);
    array.recycle();
  }

  private void setSpanCount() {
    int spanCount = Math.max(1, getMeasuredWidth() / columnWidth);
    manager.setSpanCount(spanCount);
  }
}

File: app/src/main/kotlin/com/github/mbarrben/moviedb/movies/view/MoviesRecyclerView.kt
package com.github.mbarrben.moviedb.movies.view

import android.content.Context
import android.support.v7.widget.GridLayoutManager
import android.support.v7.widget.RecyclerView
import android.util.AttributeSet
import com.github.mbarrben.moviedb.extensions.use

class MoviesRecyclerView(context: Context, attrs: AttributeSet) : RecyclerView(context, attrs) {

  private val manager: GridLayoutManager
  private var columnWidth = -1

  init {
    setHasFixedSize(true)
    addItemDecoration(MovieItemDecoration(getContext()))
    initColumnsWidth(context, attrs)
    manager = GridLayoutManager(getContext(), 1)
    layoutManager = manager
  }

  override fun onMeasure(widthSpec: Int, heightSpec: Int) {
    super.onMeasure(widthSpec, heightSpec)

    if (columnWidth > 0) {
      setSpanCount()
    }
  }

  private fun initColumnsWidth(context: Context, attrs: AttributeSet?) {
    checkNotNull(attrs) { "attrs can't be null" }

    val attrsArray = intArrayOf(android.R.attr.columnWidth)

    context.obtainStyledAttributes(attrs, attrsArray).use {
      columnWidth = getDimensionPixelSize(0, -1)
    }
  }

  private fun setSpanCount() {
    val spanCount = Math.max(1, measuredWidth / columnWidth)
    manager.spanCount = spanCount
  }
}
